server/pkcs11.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// PKCS11URI defines a PKCS#11 URI: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct { // pkcs11:path-attr[?query-attr]
+	// path-attr: (; delimited)
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS (cert, data, private, public, or secret-key)
+	Id     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotId      uint //           slot-id <- CK_SLOT_ID
+
+	// query-attr: (& delimited)
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+}
+
+func RFC7512Parser(uri string) PKCS11URI {
+	aChar := "[a-z-_]"
+	pChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=&]"
+	pAttr := aChar+"+="+pChar+"+"
+	pClause := "("+pAttr+";)*("+pAttr+")"
+	qChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=/\\?\\|]"
+	qAttr := aChar+"+="+qChar+"+"
+	qClause := "("+qAttr+"&)*("+qAttr+")"
+	r, _ := regexp.Compile("^pkcs11:"+pClause+"(\\?"+qClause+")?$")
+
+	if ! r.MatchString(uri) {
+		log.Error("PKCS#11 URI is malformed: " + uri)
+		return PKCS11URI{Id: []byte{}}
+	}
+
+	var pk11uri PKCS11URI
+	var pAttrs []string
+	var qAttrs []string
+	var parts  []string
+	var value string
+
+	uri = strings.Split(uri, "pkcs11:")[1]
+	parts = strings.Split(uri, "?")
+	pAttrs = strings.Split(parts[0], ";")
+	if 1 < len(parts) {
+		qAttrs = strings.Split(parts[1], "&")
+	}
+
+	for _, attr := range pAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if 1 < len(parts) {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		}
+		switch parts[0] {
+		case "token":
+			pk11uri.Token = value
+		case "manufacturer":
+			pk11uri.Manuf = value
+		case "serial":
+			pk11uri.Serial = value
+		case "model":
+			pk11uri.Model = value
+		case "library-manufacturer":
+			pk11uri.LibManuf = value
+		case "library-description":
+			pk11uri.LibDesc = value
+		case "library-version":
+			pk11uri.LibVer = value // TODO 1*DIGIT [ "." 1*DIGIT ]
+		case "object":
+			pk11uri.Object = []byte(value)
+		case "type":
+			pk11uri.Type = value // TODO public, private, cert, secret-key, data
+		case "id":
+			pk11uri.Id = []byte(value)
+		case "slot-manufacturer":
+			pk11uri.SlotManuf = value
+		case "slot-description":
+			pk11uri.SlotDesc = value
+		case "slot-id":
+			id, _ := strconv.ParseUint(value, 10, 32) // TODO what is the bit size
+			pk11uri.SlotId = uint(id)
+		default:
+			log.Warning("Unrecognized PKCS#11 URI Component: " + parts[0])
```

Comment:
Instead of logging an error, let's return one here. It's best to fail loudly when introducing new logic as we can more easily catch problems.